### PR TITLE
Device manager - prune client information events after remote sign out

### PR DESCRIPTION
**Pull request by kerryarchibald**
## Checklist

-   [ ] Tests written for new code (and old code if feasible)
-   [ ] Linter and other CI checks pass
-   [ ] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

---
Here's what your changelog entry will look like:

## ✨ Features
 * Device manager - prune client information events after remote sign out ([\#9874](https://github.com/matrix-org/matrix-react-sdk/pull/9874)). Contributed by @kerryarchibald.

### CODE REVIEW

File: src/components/views/settings/devices/useOwnDevices.ts [hunk 1]
```diff
@@ -116,8 +116,8 @@ export type DevicesState = {
 export const useOwnDevices = (): DevicesState => {
     const matrixClient = useContext(MatrixClientContext);
 
-    const currentDeviceId = matrixClient.getDeviceId();
-    const userId = matrixClient.getUserId();
+    const currentDeviceId = matrixClient.getDeviceId()!;
```
Comment (weeman1337) [inline src/components/views/settings/devices/useOwnDevices.ts:119]: We maybe need `getSafeDeviceId()` :upside_down_face:

File: src/utils/device/clientInformation.ts [hunk 2]
```diff
@@ -66,12 +66,29 @@ export const recordClientInformation = async (
...
 
 /**
- * Remove extra client information
- * @todo(kerrya) revisit after MSC3391: account data deletion is done
- * (PSBE-12)
+ * Remove client information events for devices that no longer exist
+ * @param validDeviceIds - ids of current devices,
+ *                      client information for devices NOT in this list will be removed
+ */
+export const pruneClientInformation = async (validDeviceIds: string[], matrixClient: MatrixClient): Promise<void> => {
```
Comment (weeman1337) [inline src/utils/device/clientInformation.ts:73]: I can find neither an `await` nor a Promise return here. Should it really be `async`?

File: src/components/views/settings/devices/useOwnDevices.ts [hunk 3]
```diff
@@ -176,6 +171,15 @@ export const useOwnDevices = (): DevicesState => {
         refreshDevices();
     }, [refreshDevices]);
 
+    useEffect(() => {
+        const deviceIds = Object.keys(devices);
+        // empty devices means devices have not been fetched yet
+        // as there is always at least the current device
+        if (deviceIds.length) {
+            pruneClientInformation(Object.keys(devices), matrixClient);
```
Comment (weeman1337) [inline src/components/views/settings/devices/useOwnDevices.ts:179]: One `Object.keys` call can be saved here:

```suggestion
            pruneClientInformation(deviceIds, matrixClient);
```

File: src/utils/device/clientInformation.ts [hunk 4]
```diff
@@ -66,12 +66,29 @@ export const recordClientInformation = async (
...
+    const orphanClientInformationEvents = Object.values(matrixClient.store.accountData).filter((event) => {
+        if (event.getType().startsWith(clientInformationEventPrefix)) {
+            const [, deviceId] = event.getType().split(clientInformationEventPrefix);
+            return deviceId && !validDeviceIds.includes(deviceId);
```
Comment (weeman1337) [inline src/utils/device/clientInformation.ts:77]: What about calling `deleteAccountData` here directly?

**Review by weeman1337 — COMMENTED** [review]
Looks good in general :+1: 

Some smaller notes/questions

**Review by SimonBrandner — APPROVED** [review]
Looks sane

**Pull request merged**